base/sparse/sparsevector.jl

```diff
@@ -0,0 +1,245 @@
+        numnz = L-ndups
+        deleteat!(I, (numnz+1):L)
+        deleteat!(V, (numnz+1):length(V))
+    end
+
+    return SparseVector(m, I, V)
+end
+
+function Base.convert{Tv,Ti}(::Type{SparseVector{Tv,Ti}}, V::Vector)
+    m = length(V)
+    I, V = findnz(V)
+    return sparse_I_sorted!(convert(Vector{Ti},I),
+                            convert(Vector{Tv},V),
+                             m)
+end
+Base.convert{T}(::Type{AbstractVector{T}}, A::SparseVector) = convert(SparseVector{T}, A)
+Base.convert(::Type{Vector}, S::SparseVector) = full(S)
+
+function Base.full{Tv}(S::SparseVector{Tv})
+    A = zeros(Tv, S.m)
+    for m = 1:length(S.rowval)
+        A[S.rowval[m]] = S.nzval[m]
+    end
+    return A
+end
+
+Base.float(S::SparseVector) = SparseVector(S.m, copy(S.rowval), float(copy(S.nzval)))
+
+Base.complex(S::SparseVector) = SparseVector(S.m, copy(S.rowval), complex(copy(S.nzval)))
+
+Base.complex(A::SparseVector, B::SparseVector) = A + im*B
+
+#TODO: remove other definition in sparsematrix.jl
+#=function Base.vec{Tv,Ti}(S::SparseMatrixCSC{Tv,Ti})
+    colptr = Array(Ti,2)
+    rowval = similar(S.rowval)
+    lS = length(S)
+    sparse_compute_reshaped_colptr_and_rowval(colptr, rowval, lS, 1, S.colptr, S.rowval, S.m, S.n)
+    return SparseVector{Tv,Ti}(lS, rowval, copy(S.nzval))
+end=#
+
+#sparsevec(A::AbstractMatrix) = reshape(sparse(A), (length(A),1))
+sparsevec(S::SparseMatrixCSC) = vec(S)
+
+sparsevec{K<:Integer,V}(d::Dict{K,V}, len::Int) = sparsevec(collect(keys(d)), collect(values(d)), len)
+
+sparsevec{K<:Integer,V}(d::Dict{K,V}) = sparsevec(collect(keys(d)), collect(values(d)))
+
+sparsevec(I::AbstractVector, V, m::Integer) = sparsevec(I, V, m, +)
+
+sparsevec(I::AbstractVector, V) = sparsevec(I, V, maximum(I), +)
+
+function sparsevec(I::AbstractVector, V, m::Integer, combine::Function)
+    nI = length(I)
+    if isa(V, Number); V = fill(V, nI); end
+    p = sortperm(I)
+    @inbounds I = I[p]
+    (nI==0 || m >= I[end]) || throw(DimensionMismatch("indices cannot be larger than length of vector"))
+    (nI==0 || I[1] > 0) || throw(BoundsError())
+    V = V[p]
+    sparse_I_sorted!(I, V, m, combine)
+end
+
+function sparsevec(V::Vector)
+    m = length(V)
+    I, V = findnz(V)
+    return sparse_I_sorted!(I,V,m,+)
+end
+
+sparse(a::Vector) = sparsevec(a)
+
+function sprand{T}(m::Integer, density::FloatingPoint,
+                   rng::Function,::Type{T}=eltype(rng(1)))
+    0 <= density <= 1 || throw(ArgumentError("$density not in [0,1]"))
+    m == 0 && return spzeros(T,m)
+    # TODO: need to change rng(1)?
+    m == 1 && return rand() <= density ? sparse(rng(1)) : spzeros(T,1)
+
+    I = Array(Int, 0) # indices of nonzero elements
+    sizehint(I, int(N*density))
+
+    # density of nonzero columns:
+    L = log1p(-density)
+    coldensity = -expm1(m*L) # = 1 - (1-density)^m 
+    colsparsity = exp(m*L) # = 1 - coldensity
+    L = 1/L
+    
+    rows = Array(Int, 0)
+    for j in randsubseq(1:n, coldensity)
```

Comment:
This logic is obviously for multiple-column cases. For a single vector, it can be much simpler.